### PR TITLE
added API for ql as datasource

### PR DESCRIPTION
**Pull request by pooja1pathak**
## Proposed changes

Added a new API ```/v2/<entityId>``` for ql to act as datasource. Fix for issue #102 

## Types of changes

What types of changes does your code introduce to the project: _Put
an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to
ask. We're here to help! This is simply a reminder of what we are going
to look for before merging your code._

- [x] I have read the [CONTRIBUTING][contrib] doc
- [x] I have signed the [CLA][cla]
- [x] I have added tests that prove my fix is effective or that my
      feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in
      downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion
by explaining why you chose the solution you did and what alternatives
you considered, etc...

[cla]: https://raw.githubusercontent.com/orchestracities/ngsi-timeseries-api/master/individual-cla.pdf
    "Martel Open Source Software Individual Contributor License Agreement"
[contrib]: https://github.com/smartsdk/ngsi-timeseries-api/blob/master/CONTRIBUTING.md
    "Contributing to QuantumLeap"

### CODE REVIEW

**Comment by chicco785** [comment]
1. rather than having a new end point, the endpoint should be `/v2/entities/{entityId}` the new behaviour could be triggered by a query parameter (e.g. `?datasource=true`)
2. the returned format should be the same as the one of ngsi entity not the timeseries version of it, e.g.:
```
{
      "id": "Kitchen1",
      "type": "Kitchen",
      "temperature": {
        "value": 35.6,
        "type": "Number"
      },
     "dateModified": "value from time index"
  }
```
3. the returned value, should be the last value based on the time_index

**Comment by pooja1pathak** [comment]
Hi @chicco785 I have updated the code as per your comment. Please check and let me know if any further change is required.

File: src/reporter/tests/test_1T1ENA.py [hunk 1]
```diff
@@ -632,3 +632,33 @@ def test_not_found(service):
...
+    h = {'Fiware-Service': service}
+
+    r = requests.get(query_url(), params=query_params, headers=h)
+    assert r.status_code == 200, r.text
+
+    # Assert
+    expected = {
+        'id': entity_id,
+        'type': entity_type,
+        'index': ['1970-01-30T00:00:00+00:00'],
```
Comment (chicco785) [inline src/reporter/tests/test_1T1ENA.py:651]: this should be `modifiedAt` and the value should be just the first element of the array, not the full array.

Comment (chicco785) [reply src/reporter/tests/test_1T1ENA.py:651]: this means, you need to manipulate the result when `datasource=true`

**Review by chicco785 — CHANGES_REQUESTED** [review]

File: src/reporter/tests/test_1T1ENA.py (hunk 1 above)
Comment (pooja1pathak) [reply src/reporter/tests/test_1T1ENA.py]: @chicco785  I have updated the code as per your comment

**Comment by pooja1pathak** [comment]
> to complete the testing, we should add in the integration test something that:
> 
> 1. push an entity into quantumleap
> 2. register the datasource endpoint for that entity in orion
> 3. check that orion can retrieve the entity via ql

I am working on this. I will update PR once changes are done.

File: src/reporter/tests/test_1T1ENA.py (hunk 1 above)
Comment (chicco785) [reply src/reporter/tests/test_1T1ENA.py]: LGTM!

**Comment by pooja1pathak** [comment]
Hi @chicco785 

I have tried to add integration test which you have suggested. Please find my below findings:

I have registered the ql endpoint in orion as datasource using API:
```
(curl localhost:1026/v1/registry/registerContext -s -S --header 'Content-Type: application/json' \
    --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
{
    "contextRegistrations": [
        {
            "entities": [
                {
                    "type": "Test",
                    "isPattern": "false",
                    "id": "Test1"
                }
            ],
            "attributes": [
                {
                    "name": "temperature",
                    "type": "integer"
                }
            ],
            "providingApplication": "http://localhost:8668/v2/entities/Test1%3Fdatasource%3Dtrue"
        }
    ],
    "duration": "P1M"
}
EOF
```
%3F and %3D is encoded character for ? and = respectively.
But when I send query request to orion as:
```
(curl localhost:1026/v1/queryContext -s -S --header 'Content-Type: application/json' \
    --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
{
    "entities": [
        {
            "type": "Test",
            "isPattern": "false",
            "id": "Test1"
        }
    ],
    "attributes": [
    ]
}
EOF
```
It returned:
```
{
    "errorCode": {
        "code": "404",
        "details": "invalid context provider response",
        "reasonPhrase": "No context element found"
    }
}
```
When I checked orion logs, I found:
```
 POST http://localhost:8668/v2/entities/Test1%3Fdatasource%3Dtrue/queryContext, request payload (114 bytes): {"entities":[{"type":"Test","isPattern":"false","id":"Test1"}],"attributes":["temperature"]}, response payload (0 bytes): , response code: Couldn't connect to server
```
As per my observation Orion add ```/queryContext``` to the url in ```providingApplication```.
This may be the issue with Orion. We should clarify this orion community.

**Comment by chicco785** [comment]
> Hi @chicco785
> 
> I have tried to add integration test which you have suggested. Please find my below findings:
> 
> I have registered the ql endpoint in orion as datasource using API:
> 
> ```
> (curl localhost:1026/v1/registry/registerContext -s -S --header 'Content-Type: application/json' \
>     --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
> {
>     "contextRegistrations": [
>         {
>             "entities": [
>                 {
>                     "type": "Test",
>                     "isPattern": "false",
>                     "id": "Test1"
>                 }
>             ],
>             "attributes": [
>                 {
>                     "name": "temperature",
>                     "type": "integer"
>                 }
>             ],
>             "providingApplication": "http://localhost:8668/v2/entities/Test1%3Fdatasource%3Dtrue"
>         }
>     ],
>     "duration": "P1M"
> }
> EOF
> ```
> 
> %3F and %3D is encoded character for ? and = respectively.
> But when I send query request to orion as:
> 
> ```
> (curl localhost:1026/v1/queryContext -s -S --header 'Content-Type: application/json' \
>     --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
> {
>     "entities": [
>         {
>             "type": "Test",
>             "isPattern": "false",
>             "id": "Test1"
>         }
>     ],
>     "attributes": [
>     ]
> }
> EOF
> ```
> 
> It returned:
> 
> ```
> {
>     "errorCode": {
>         "code": "404",
>         "details": "invalid context provider response",
>         "reasonPhrase": "No context element found"
>     }
> }
> ```
> 
> When I checked orion logs, I found:
> 
> ```
>  POST http://localhost:8668/v2/entities/Test1%3Fdatasource%3Dtrue/queryContext, request payload (114 bytes): {"entities":[{"type":"Test","isPattern":"false","id":"Test1"}],"attributes":["temperature"]}, response payload (0 bytes): , response code: Couldn't connect to server
> ```
> 
> As per my observation Orion add `/queryContext` to the url in `providingApplication`.
> This may be the issue with Orion. We should clarify this with orion community.

not sure this is due to the fact that you tested v1 registration and hence it expects v1 data format, maybe try with v2 api:

https://fiware-orion.readthedocs.io/en/1.15.0/user/walkthrough_apiv2/index.html#context-availability-management

otherwise, something in the format we use is not correct :(

**Comment by chicco785** [comment]
apparently, is a bit more complex that what we thought, here you can find an example of a context provider: https://github.com/FIWARE/tutorials.NGSI-v2/tree/master/context-provider

basically we would need a whole new endpoint: `/op/query` (as post, not get), and this endpoint would need to accept requests in the format of NGSIv2 `/op/query` endpoint:
https://fiware.github.io/specifications/ngsiv2/stable/

i.e.: something like:
```
{
  "entities": [
    {
      "idPattern": ".*",
      "type": "Room"
    },
    {
      "id": "Car",
      "type": "P-9873-K"
    }
  ],
  "attrs": [
    "temperature",
    "humidity"
  ],
  "expression": {
    "q": "temperature>20"
  },
  "metadata": [
    "accuracy",
    "timestamp"
  ]
}
```

In our case we can skip expression (we don't support them) and metadata (we don't store them). Moreover, as initial implementation, we can disregard the usage of patterns and support only explicit id and type. This should make it simple enough to implement it starting from what you coded so far. Feelings?

**Comment by pooja1pathak** [comment]
sure, First I will confirm ```orion v2/registrations``` behaviour then update PR as per your suggestion.

**Comment by pooja1pathak** [comment]
@chicco785 I checked with orion ```/v2/registrations```, I found that orion append ```/queryContext``` to the ```provider url```. 
So we need to create a new API as ```/v2/op/queryContext``` which will respond to POST request received from orion. I am working on it and update the PR once changes are done.

**Comment by chicco785** [comment]
this function I developed may be of help:
https://github.com/orchestracities/ngsi-timeseries-api/blob/master/src/translators/sql_translator.py#L1205

it also leverages the modified `format_response` function now is able to format messages for single point in time value entities:
https://github.com/orchestracities/ngsi-timeseries-api/blob/master/src/translators/sql_translator.py#L1415

**Comment by pooja1pathak** [comment]
@chicco785 I have updated the code as per our discussion. Please verify if it ok then I'll add test cases for the same.
I have executed below request:
```
(curl localhost:8668/v2/op/queryContext -s -S --header 'Content-Type: application/json' \
    --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
{
    "entities": [
        {
            "type": "TestNew",
            "isPattern": "false",
            "id": "TestNew1"
        },
        {
            "type": "Test",
            "isPattern": "false",
            "id": "Test1"
        }
    ],
    "attributes": [
        "temperature",
        "intensity"
    ]
}
EOF
```
and received response as:
```
[
    [
        {
            "dateModified": "2021-08-24T19:57:22.731+00:00",
            "id": "TestNew1",
            "intensity": {
                "type": "Text",
                "value": "Low"
            },
            "temperature": {
                "type": "Number",
                "value": 22.0
            },
            "type": "TestNew"
        }
    ],
    [
        {
            "dateModified": "2021-08-24T19:55:15.701+00:00",
            "id": "Test1",
            "intensity": {
                "type": "Text",
                "value": "Low"
            },
            "temperature": {
                "type": "Number",
                "value": 22.0
            },
            "type": "Test"
        }
    ]
]
```

Please let me know if any further change is required.

File: src/reporter/queryContext.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            }, 422
+
+        except Exception as e:
+            # Temp workaround to debug test_not_found
+            msg = "Something went wrong with QL. Error: {}".format(e)
+            logging.getLogger(__name__).error(msg, exc_info=True)
+            return msg, 500
+
+        if entities:
+            if len(entities) > 1:
```
Comment (chicco785) [inline src/reporter/queryContext.py:68]: be careful, it may be more than one entity.

**Comment by chicco785** [comment]
is `query` or `queryContext`, according to the docs it should be `query` in ngsi v2, while it's `queryContext` in ngsi v1 (see spec: https://fiware.github.io/specifications/ngsiv2/stable/ ) but implementation may be different...

also the input message according to ngsi v2 spec should be like:

```
{
  "entities": [
    {
      "id": "TestNew1",
      "type": "TestNew"
    },
    {
      "id": "Test1",
      "type": "Test"
    }
  ],
  "attrs": [
    "temperature",
    "intensity"
  ]
}
```

also, when doing validation, if the posted request includes `idPattern` or `expression` or `metadata`, we should return a message "not supported", e.g.:


```
{
  "entities": [
    {
      "idPattern": ".*",
      "type": "Room"
    }
  ],
  "expression": {
    "q": "temperature>20"
  },
  "metadata": [
    "accuracy",
    "timestamp"
  ]
}
```


> @chicco785 I have updated the code as per our discussion. Please verify if it ok then I'll add test cases for the same.
> I have executed below request:
> 
> ```
> (curl localhost:8668/v2/op/queryContext -s -S --header 'Content-Type: application/json' \
>     --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
> {
>     "entities": [
>         {
>             "type": "TestNew",
>             "isPattern": "false",
>             "id": "TestNew1"
>         },
>         {
>             "type": "Test",
>             "isPattern": "false",
>             "id": "Test1"
>         }
>     ],
>     "attributes": [
>         "temperature",
>         "intensity"
>     ]
> }
> EOF
> ```
> 
> and received response as:
> 
> ```
> [
>     [
>         {
>             "dateModified": "2021-08-24T19:57:22.731+00:00",
>             "id": "TestNew1",
>             "intensity": {
>                 "type": "Text",
>                 "value": "Low"
>             },
>             "temperature": {
>                 "type": "Number",
>                 "value": 22.0
>             },
>             "type": "TestNew"
>         }
>     ],
>     [
>         {
>             "dateModified": "2021-08-24T19:55:15.701+00:00",
>             "id": "Test1",
>             "intensity": {
>                 "type": "Text",
>                 "value": "Low"
>             },
>             "temperature": {
>                 "type": "Number",
>                 "value": 22.0
>             },
>             "type": "Test"
>         }
>     ]
> ]
> ```
> 
> Please let me know if any further change is required.

File: src/reporter/queryContext.py (hunk 2 above)
Comment (pooja1pathak) [reply src/reporter/queryContext.py:68]: As per my understanding for 1 entityID and entityType there should be only one entity so only one entity will be returned by the method. Please correct me if I am wrong.

Comment (chicco785) [reply src/reporter/queryContext.py:68]: you are right, they become more than 1 only after appending them in `res`!

File: src/reporter/reporter.py [hunk 3]
```diff
@@ -106,6 +106,26 @@ def _validate_payload(payload):
                 'An entity update is missing value '
                 'for attribute {}'.format(attr))
 
+def _validate_body(payload):
```
[the remark this answers is not in the document]
Comment (pooja1pathak) [reply src/reporter/reporter.py]: done

**Comment by pooja1pathak** [comment]
I have done the changes as discussed. Please check and let me know if any further change is required.

File: specification/quantumleap.yml [hunk 4]
```diff
@@ -138,7 +138,7 @@ definitions:
         $ref: '#/definitions/ValuesArray'
 
   AttributesArray:
-    type: array
+    type: string
```
Comment (chicco785) [inline specification/quantumleap.yml:141]: array is correct

Comment (pooja1pathak) [reply specification/quantumleap.yml:141]: should I update it now or leave?

Comment (chicco785) [reply specification/quantumleap.yml:141]: #547 will handle it ;)

**Comment by chicco785** [comment]
@pooja1pathak thanks for this valuable contribution!

**Pull request merged**